industrialucn/optimization/test_econstraint.py

```diff
@@ -0,0 +1,290 @@
+
+        v2 = [[1.0, 4324.8, 165.4],
+              [2.0, 2757.5, 116.1],
+              [3.0, 2106.9, 74.8],
+              [13.0, 1351.1, 56.9],
+              [2.0, 2757.5, 116.1],
+              [2.0, 2966.0, 107.7],
+              [3.0, 2106.9, 74.8],
+              [13.0, 1351.1, 56.9],
+              [3.0, 2223.9, 82.5]]
+
+        v3 = [[1.0, 4324.8, 165.4],
+              [2.0, 2757.5, 116.1],
+              [3.0, 2106.9, 74.8],
+              [13.0, 1351.1, 56.9],
+              [2.0, 2757.5, 116.1],
+              [2.0, 2966.0, 107.7],
+              [3.0, 2106.9, 74.8],
+              [13.0, 1351.1, 56.9],
+              [3.0, 2223.9, 82.5],
+              [2.0, 2106.9, 74.8]]
+
+        self.assertTrue(all_non_dominated(v1))
+        self.assertFalse(all_non_dominated(v2))
+        self.assertFalse(all_non_dominated(v3))
+
+        self.assertTrue(all_weakly_non_dominated(v1))
+        self.assertTrue(all_weakly_non_dominated(v2))
+        self.assertFalse(all_weakly_non_dominated(v3))
+
+    def test_cplex_mip(self):
+        from docplex.mp.model import Model as CplexModel
+
+        config_logger(ec.logger, logging.DEBUG)
+
+        # define model
+        mdl = CplexModel()
+
+        # define variables
+        x = mdl.binary_var_dict(parks, name='x')
+        y = mdl.binary_var_dict([(i, j) for i in buildings for j in parks], name='y')
+
+        # define objective
+        f1 = -mdl.sum(x[i] for i in parks)
+        f2 = -mdl.sum(distance[i, j] * y[i, j] for i in buildings for j in parks)
+        f3 = -mdl.max(distance[i, j] * y[i, j] for i in buildings for j in parks)
+        objectives = [f1, f2, f3]
+
+        # define constraints
+        mdl.add_constraints(y[i, j] <= x[j] for i in buildings for j in parks)
+        mdl.add_constraints(mdl.sum(y[i, j] for j in parks) >= 1 for i in buildings)
+
+        pareto_frontier = []
+
+        def extract_solution():
+            xvals = {i: x[i].solution_value for i in parks}
+            yvals = {(i, j): y[i, j].solution_value for i in buildings for j in parks}
+            pareto_frontier.append((xvals, yvals))
+
+        # run e-constraint
+        ec.run_econstraint(model=mdl,
+                           objectives=objectives,
+                           solution_extractor=extract_solution,
+                           optimizer='cplex')
+
+        def f(i, sol):
+            x, y = sol
+            if i == 0:
+                return sum(x[i] for i in parks)
+            elif i == 1:
+                return sum(distance[i, j] * y[i, j] for i in buildings for j in parks)
+            else:
+                return max(distance[i, j] * y[i, j] for i in buildings for j in parks)
+
+        fvs = [[f(i, sol) for i in range(3)] for sol in pareto_frontier]
+
+        for fv in fvs:
+            a, b, c = fv
+            print(f'[{a:>5.1f}, {b:>7.1f}, {c:>5.1f}],')
+
+        self.assertTrue(all_weakly_non_dominated(fvs))
+
+    def test_cplex_lp(self):
```

Comment:
the name of the function needs more details

Reply:
Sure!
I made the change here 0b2563e96ded21f60427200f26f93a164b70849e